File: server/routes/ancestry.ts
/**
 * DNA-to-culture ancestry mapping routes (US-001) — **ported to Python**
 * (pinakes:65 US-2).
 *
 * `GET /api/ancestry/haplogroups` and `POST /api/ancestry/map` are served by
 * `services/api/src/pinakes/routers/ancestry.py` over
 * `pinakes.analytics.genetic`, which now carries both halves of
 * `services/genetic-linguistic-correlation.ts` and so shares one
 * `NOTABLE_DIVERGENCES` table between the mapper and the correlation engine —
 * the thing pinakes:62 US-1 left a note asking for. Both handlers here answer
 * **501** naming their replacement.
 *
 * **The privacy guarantee is unaffected, because it was never here.** Raw-DNA
 * parsing and haplogroup inference happen in the browser (`web/src/lib/dna/*`);
 * only non-identifying haplogroup ids ever left it, and the port moved the
 * enrichment step alone.
 *
 * `services/genetic-linguistic-correlation.ts` is **not** retired — it is the
 * graded spec, and `services/ancestry-mapper.test.ts` is what says the two
 * mappers agree, including the bare-name → namespaced-id fallback without which
 * the map is empty against live data.
 */
import type { Express, Request, Response } from "express";

/** The Python module that owns these routes now. */
export const PORTED_TO = "services/api/src/pinakes/routers/ancestry.py";

/** Machine-readable discriminator in a retired route's body. */
export const PORTED_ERROR = "ported";

/** The routes this file no longer serves. */
export const PORTED_ROUTES = {
  haplogroups: "/api/ancestry/haplogroups",
  map: "/api/ancestry/map",
} as const;

/**
 * A handler for a route this backend no longer owns.
 *
 * 501, not 404 or 503: the route still exists in the API contract and something
 * does serve it — just not this process.
 */
function portedToPython(route: string) {
  return (_req: Request, res: Response): void => {
    res.status(501).json({
      error: PORTED_ERROR,
      message:
        `${route} has been ported to the Python service and is served there ` +
        `(${PORTED_TO}). The Express handler is retired.`,
      route,
      servedBy: PORTED_TO,
      coverage: "/api/_parity/coverage",
    });
  };
}

export function registerAncestryRoutes(app: Express): void {
  app.get(
    PORTED_ROUTES.haplogroups,
    portedToPython(`GET ${PORTED_ROUTES.haplogroups}`),
  );
  app.post(PORTED_ROUTES.map, portedToPython(`POST ${PORTED_ROUTES.map}`));
}

File: services/api/src/pinakes/routers/ancestry.py
"""The `/api/ancestry/*` route group — DNA to cultural association.

Ported off `server/routes/ancestry.ts` (pinakes:65 US-2) over
:mod:`pinakes.analytics.genetic`, which already owned the divergence table this
mapper shares with the correlation engine.

**The privacy guarantee is upstream of this file and stays there.** Raw-DNA
parsing and haplogroup inference happen in the browser (`web/src/lib/dna/*`); a
request carries only the inferred haplogroup ids — never genotypes — and all
these two handlers do is enrich non-identifying ids from the public reference
corpus. Nothing here reads, writes or logs a request body beyond the id list, and
that is deliberate rather than incidental.

The corpus haplogroups are **Y-chromosome only**, so an association is
paternal-line only. The result says so itself: every answer carries
:data:`~pinakes.analytics.genetic.ANCESTRY_CAVEATS`, including the answer that
matched nothing.
"""

from __future__ import annotations

import logging
from typing import Annotated, Any

from fastapi import APIRouter, Body
from fastapi.responses import JSONResponse

from pinakes.analytics import genetic
from pinakes.paths import lexicons_dir

logger = logging.getLogger("pinakes.ancestry")

router = APIRouter(tags=["ancestry"])


@router.get("/api/ancestry/haplogroups")
def haplogroups() -> Any:
    """The reference haplogroups the mapper recognizes."""
    try:
        return genetic.reference_haplogroups(genetic.load_ancestry_data(lexicons_dir()))
    except Exception as error:  # noqa: BLE001 - reported as the Express 500
        logger.exception("failed to load haplogroups")
        return JSONResponse(
            status_code=500,
            content={"error": "failed to load haplogroups", "detail": str(error)},
        )


@router.post("/api/ancestry/map")
def map_ancestry(body: Annotated[Any, Body()] = None) -> Any:
    """Map inferred haplogroup ids to cultural associations.

    Two 400s rather than one, and they say different things: a missing or empty
    list is a malformed request, where a list of non-strings is a request whose
    ids cannot be read. Express distinguished them and the client surfaces the
    message verbatim.
    """
    payload = body if isinstance(body, dict) else {}
    raw = payload.get("haplogroupIds")
    if not isinstance(raw, list) or not raw:
        return JSONResponse(
            status_code=400,
            content={"error": "haplogroupIds must be a non-empty array of strings"},
        )
    haplogroup_ids = [item for item in raw if isinstance(item, str)]
    if not haplogroup_ids:
        return JSONResponse(
            status_code=400,
            content={"error": "haplogroupIds must contain at least one string id"},
        )

    try:
        return genetic.map_haplogroups_to_ancestry(
            haplogroup_ids, genetic.load_ancestry_data(lexicons_dir())
        )
    except Exception as error:  # noqa: BLE001 - reported as the Express 500
        logger.exception("Unexpected error in /api/ancestry/map")
        return JSONResponse(
            status_code=500,
            content={"error": "ancestry mapping failed", "detail": str(error)},
        )
